### PR TITLE
Added optimizable gmm fisher vector, column pca estimators

### PR DESCRIPTION
**Pull request by tomerk**
issue #239

### CODE REVIEW

File: src/main/scala/nodes/images/FisherVector.scala [hunk 1]
```diff
@@ -49,3 +51,45 @@ case class FisherVector(gmm: GaussianMixtureModel)
...
+ * Trains either a scala or an `enceval` Fisher Vector implementation, via
+ * estimating a GMM by treating each column of the inputs as a separate
+ * DenseVector input to [[GaussianMixtureModelEstimator]]
+ *
+ * Automatically decides which implementation to use when node-level optimization is enabled.
+ *
+ * @param k Number of centers to estimate.
+ */
+
+case class OptimizableGMMFisherVectorEstimator(k: Int) extends OptimizableEstimator[DenseMatrix[Float], DenseMatrix[Float]] {
```
Comment (etrain) [inline src/main/scala/nodes/images/FisherVector.scala]: Can we not call these things `Optimizable` when they are? I really don't want users to have to think about what's going on under the covers - the fact that the system is optimizing this under the covers should be incidental to their experience.

Comment (etrain) [inline src/main/scala/nodes/images/FisherVector.scala]: Even if this means changing the name of the old one!

File: src/main/scala/nodes/learning/PCA.scala [hunk 2]
```diff
@@ -55,6 +56,44 @@ case class ColumnPCAEstimator(dims: Int) extends Estimator[DenseMatrix[Float], D
...
+/**
+ * Estimates a PCA model for dimensionality reduction based on a sample of a larger input dataset.
+ * Treats each column of the input matrices like a separate DenseVector input to [[PCAEstimator]] or
+ * [[DistributedPCAEstimator]].
+ *
+ * Automatically decides between distributed and local implementations when node-level optimization is enabled.
+ *
+ * @param dims Dimensions to reduce input dataset to.
+ */
+class OptimizableColumnPCAEstimator(dims: Int) extends OptimizableEstimator[DenseMatrix[Float], DenseMatrix[Float]] {
```
Comment (etrain) [inline src/main/scala/nodes/learning/PCA.scala]: Same thing here.

File: src/main/scala/nodes/learning/PCA.scala [hunk 3]
```diff
@@ -55,6 +56,44 @@ case class ColumnPCAEstimator(dims: Int) extends Estimator[DenseMatrix[Float], D
...
+  val default = new DistributedColumnPCAEstimator(dims)
+
+  override def optimize(sample: RDD[DenseMatrix[Float]], numPerPartition: Map[Int, Int])
+  : RDD[DenseMatrix[Float]] => Pipeline[DenseMatrix[Float], DenseMatrix[Float]] = {
+    val numFeatures = sample.first().rows
+
+    if (numFeatures <= 100000) {
```
Comment (etrain) [inline src/main/scala/nodes/learning/PCA.scala]: I don't think this should be "numFeatures" - should instead be `sample.count()`

**Comment by tomerk** [comment]
Okay @etrain @shivaram

File: src/main/scala/nodes/learning/PCA.scala [hunk 4]
```diff
@@ -47,11 +48,111 @@ case class BatchPCATransformer(pcaMat: DenseMatrix[Float]) extends Transformer[D
...
+
+/**
+ * Estimates a PCA model for dimensionality reduction based on a sample of a larger input dataset,
+ * using a distributed PCA algorithm.
+ * Treats each column of the input matrices like a separate DenseVector input to [[DistributedPCAEstimator]].
+ *
+ * @param dims Dimensions to reduce input dataset to.
+ */
+case class DistributedColumnPCAEstimator(dims: Int) extends Estimator[DenseMatrix[Float], DenseMatrix[Float]]
+  with SolverCostModel {
```
Comment (shivaram) [inline src/main/scala/nodes/learning/PCA.scala]: Why are all these called `SolverCostModel` ?

File: src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala [hunk 5]
```diff
@@ -48,7 +48,7 @@ object ImageNetSiftLcsFV extends Serializable with Logging {
         new BatchPCATransformer(convert(csvread(new File(fname)), Float).t)
       case None =>
         val pca = sampledColumns andThen
-            (ColumnPCAEstimator(numPCADesc), trainingData)
```
Comment (shivaram) [inline src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala]: Any reason we don't want to use the optimized nodes :) ?

File: src/main/scala/nodes/learning/PCA.scala [hunk 6]
```diff
@@ -47,11 +48,111 @@ case class BatchPCATransformer(pcaMat: DenseMatrix[Float]) extends Transformer[D
...
+    numMachines: Option[Int] = None,
+    cpuWeight: Double = 3.8e-4,
+    memWeight: Double = 2.9e-1,
+    networkWeight: Double = 1.32)
+  extends OptimizableEstimator[DenseMatrix[Float], DenseMatrix[Float]] {
+  val localEstimator = new LocalColumnPCAEstimator(dims)
+  val distributedEstimator = new DistributedColumnPCAEstimator(dims)
+  val default = distributedEstimator
+
+  override def optimize(sample: RDD[DenseMatrix[Float]], numPerPartition: Map[Int, Int])
```
Comment (shivaram) [inline src/main/scala/nodes/learning/PCA.scala]: Can we add a unit test for this similar to the solvers one ?

File: src/main/scala/nodes/learning/PCA.scala (hunk 4 above)
Comment (tomerk) [inline src/main/scala/nodes/learning/PCA.scala]: I reused the same trait I used for the linear solvers. Do you have a name you prefer?

File: src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala (hunk 5 above)
Comment (tomerk) [inline src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala]: uh not particularly, I just haven't tested these w/ the optimized nodes yet and am trying to minimize risks. I guess the integration tests will try this out for us?

File: src/main/scala/nodes/learning/PCA.scala (hunk 6 above)
Comment (tomerk) [inline src/main/scala/nodes/learning/PCA.scala]: Sure

File: src/main/scala/nodes/learning/PCA.scala (hunk 4 above)
Comment (shivaram) [inline src/main/scala/nodes/learning/PCA.scala]: Just `CostModel` would suffice -- Also its an internal API so we can augment / change this in the future I guess

File: src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala (hunk 5 above)
Comment (shivaram) [inline src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala]: Yeah testing should help - More so if we look at these pipelines as examples of how to write pipelines we want to encourage use of the optimizable nodes ? cc @etrain

**Comment by tomerk** [comment]
Unit tests added.

**Comment by shivaram** [comment]
Thanks -- changes look fine. Lets wait for @etrain to also take a pass

File: src/main/scala/nodes/learning/PCA.scala [hunk 7]
```diff
@@ -75,20 +107,50 @@ case class DistributedColumnPCAEstimator(dims: Int) extends Estimator[DenseMatri
...
+ * @param cpuWeight
+ * @param memWeight
+ * @param networkWeight
  */
-class OptimizableColumnPCAEstimator(dims: Int) extends OptimizableEstimator[DenseMatrix[Float], DenseMatrix[Float]] {
-  val default = new DistributedColumnPCAEstimator(dims)
+class ColumnPCAEstimator(
+    dims: Int,
+    numMachines: Option[Int] = None,
+    cpuWeight: Double = 3.8e-4,
```
Comment (etrain) [inline src/main/scala/nodes/learning/PCA.scala]: We should have a philosophical discussion about whether we want to expose details of optimization at this level of user API, but this is fine for now.

File: src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala (hunk 5 above)
Comment (etrain) [inline src/main/scala/pipelines/images/imagenet/ImageNetSiftLcsFV.scala]: Yeah, let's just go ahead and do it - in general I think we want to move toward the higher level expression of these pipelines, which includes a bunch of benefits like showing up in integration tests (maybe) and it encourages us to eat our own dog food and really use the optimizer.

**Comment by etrain** [comment]
Would be great to get the PCAEstimator and GMMEstimators in the existing pipelines using the `Optimizable` versions - beyond that this LGTM

**Comment by tomerk** [comment]
@etrain @shivaram I've added the optimizable versions to the existing pipelines

**Comment by etrain** [comment]
LGTM - thanks @tomerk !!

**Pull request merged**